proposals/0000-implicit-import.rst

```diff
@@ -0,0 +1,124 @@
+This removes an unnecessary restriction, making the language simpler.
+
+
+Motivation
+----------
+The main purpose is to be able to directly use a package.
+This is particularly useful for single use functions, such as ``Servant.serve``.
+For example, this propose change would make the following program valid
+(assuming a cabal file with build depends and default extensions):
+
+::
+
+ module Main where
+
+ import Servant.API (Get, JSON, (:>))
+
+ type MyApi = "answer" :> Get '[JSON] Integer
+
+ server :: Servant.Server MyApi
+ server = pure 42
+
+ main :: IO ()
+ main = Network.Wai.Handler.Warp.run 8080 app
+   where
+     app :: Network.Wai.Application
+     app = Servant.serve (Data.Proxy.Proxy @MyApi) server
+
+
+The status quo is not satisfactory because it requires adding explicit imports:
+
+::
+
+ -- Without this proposal, these imports are required
+ import qualified Data.Proxy
+ import qualified Network.Wai
+ import qualified Network.Wai.Handler.Warp
+ import qualified Servant
+
+Managing such imports has the following negative effects:
+
+- It increases the file length.
+- The author needs to jump to the top of the file each time a new module is used.
+- It complicates code refactoring when moving a definition requires moving (or copying) its import.
+
+
+Proposed Change Specification
+-----------------------------
+I propose an extension to enable implicit qualified imports.
+
+When the ``ImplicitQualifiedImport`` language extension is enabled,
+unknown fully qualified symbols are gracefully resolved early in the compilation pipeline by
+injecting an implicit ``import qualified`` statement when necessary.
+The pipeline might need an extra pass to collect missing imports.
+
+GHCi already implements this feature thanks to the flag ``-fimplicit-import-qualified``,
+but the flag does not seem to work with GHC.
+Moreover the flag is not documented in ``ghc --show-options``, but it is accepted.
+Thus this proposal deprecates the ``-fimplicit-import-qualified`` flag in favor of the
+``-XImplicitQualifiedImport`` modifier, so that the behavior is consistent between GHC and GHCi.
+
+
+Examples
+--------
+Here are some example functions that are often used locally and
+it would be useful to call them without having to add an import statement:
+
+- ``Control.Concurrent.threadDelay``
+- ``Data.Char.isAlpha``
+- ``Data.Foldable.traverse_``
+- ``Data.Maybe.mapMaybe``
+- ``Data.Set.fromList``
+- ``Data.Text.pack``
+- ``Debug.Trace.trace``
+- ``System.Environment.getArgs``
+- ``Text.Printf.printf``
+
+
+Effect and Interactions
+-----------------------
+The proposed change enables using any module without requiring an import statement.
+
+Interactions with existing language or compiler features:
+
+- Fully qualified import with hidden declarations are not affected: with ``import Data.Maybe hiding (mapMaybe)``, using ``Data.Maybe.mapMaybe`` should not be valid.
```

Comment:
What about explicitly limited imports? If I write

```haskell
import Control.Monad (liftM2)
```

will `Control.Monad.ap` be in scope?

How about renamed imports? If I write

```haskell
import Control.Monad as M hiding (ap)
```

I would expect that `M.ap` would not be valid but that `Control.Monad.ap` would be. Do you concur?